src/sources/gcp_pubsub.rs

```diff
@@ -347,34 +347,33 @@ impl PubsubSource {
         config
     }
 
-    fn request_stream(&self) -> impl Stream<Item = proto::StreamingPullRequest> + 'static {
-        // This data is only allowed in the first request
-        let mut subscription = Some(self.subscription.clone());
-        let mut client_id = Some(CLIENT_ID.clone());
-
-        let ack_ids = Arc::clone(&self.ack_ids);
+    fn request_stream(
+        &self,
+        ack_ids: UnboundedReceiver<Vec<String>>,
+    ) -> impl Stream<Item = proto::StreamingPullRequest> + 'static {
+        let ack_ids = UnboundedReceiverStream::new(ack_ids);
+        let subscription = self.subscription.clone();
+        let client_id = CLIENT_ID.clone();
         let stream_ack_deadline_seconds = self.ack_deadline_seconds;
-        stream::repeat(()).then(move |()| {
-            let ack_ids = Arc::clone(&ack_ids);
-            let subscription = subscription.take().unwrap_or_default();
-            let client_id = client_id.take().unwrap_or_default();
-            async move {
-                let mut ack_ids = ack_ids.lock().await;
-                proto::StreamingPullRequest {
-                    subscription,
-                    client_id,
-                    ack_ids: std::mem::take(ack_ids.as_mut()),
-                    stream_ack_deadline_seconds,
-                    ..Default::default()
-                }
+        stream::once(async move {
+            proto::StreamingPullRequest {
+                subscription,
+                client_id,
+                stream_ack_deadline_seconds,
+                ..Default::default()
             }
         })
+        .chain(ack_ids.map(|ack_ids| proto::StreamingPullRequest {
```

Comment:
Hmmm, so what happens when Vector acknowledges all outstanding messages, and the Pub/Sub channel has no more messages to send because it's been emptied?

Reply:
As far as I understand, as a bidirectional stream client, events (streaming pull "request" and streaming pull response) flow asynchronchronously in both directions. When the server has data, it sends it async. I instrumented the code with debugs and that is the kind of behavior I have observed.

Reply:
But based on the implication in the description of this change -- that pacing the streaming pull requests will in turn pace the streaming responses -- then doesn't it follow that you have to send a streaming pull request to get the next chunk of data?

Said another way: when there's no more IDs to acknowledge and thus no further streaming pull request, does that not deadlock the source?

Reply:
1. That was not what I was trying to communicate in the description. This change improves throughput by reducing compounding no-op requests and server-side loading.
2. Bi-directional streaming gRPC is only kinda sorta a request-response framework. It needs the initial request to set it up, but after that data flows asynchronously in both directions with no strict one-to-one correspondence between requests and responses.
3. I validated this by adding a `.ready_chunks` to the acking flow and tracked sizes with a histogram. According to that data, I sent one request with acks for up to >16 incoming responses at a time. Responses still continued to flow.

Reply:
Gotcha, so in this case, the streaming pull request is used to initialize the transaction -- client ID, etc -- and also to acknowledge responses... but you don't need to _continue_ to send the requests in order for the server to keep sending you responses as more messages come into the Pub/Sub queue/channel/whatever-its-called?

Reply:
Correct, and I strongly suspect that is also the case for any other bidirectional streaming gRPC service, though I have no other examples to vouch for that.